File: cPP_java_py_js-DSA/mix-questions/secondLargest/secondLargest.cpp
#include <bits/stdc++.h>
using namespace std;

int secondLargest(vector<int> arr ){
    int ans =arr[0]; 

    int smallestNumber =INT_MAX ; 
    int secondSmallestNumber = INT_MAX ; 


    int n = arr.size();

    for (int i=0 ;i<n ;i++ ){
        int value = arr[i]; 
        if (value<smallestNumber){

            // transfer the value of smallest into second smallest
            secondSmallestNumber = smallestNumber ;
            smallestNumber = value;

        }
        else{
            if (value<secondSmallestNumber &&  value != smallestNumber){
                secondSmallestNumber = value ; 
            }
        }
    }



    ans = secondSmallestNumber ; 

    return ans;
}

void printArray(vector<int>arr ){
    int n = arr.size(); 

    for (int i=0 ;i<n ;i++ ){
        cout<<arr[i]<<" ";
    }
    cout<<endl;
}

int main()
{
    vector<int> arr = {12, 35, 1, 10, 34, 1};
    printArray(arr);
    sort(arr.begin(),arr.end());
    printArray(arr);
    cout << "second Largest element is "<< secondLargest(arr) << endl;
    return 0;
}

File: cPP_java_py_js-DSA/mix-questions/secondLargest/secondLargest.js
function secondLargest(arr) {
  // optimised approach: (using two variables)

  console.log("Time complexity:O(n");
  console.log("Space complexity:O(1)");
  let largest = Number.MIN_SAFE_INTEGER;
  let secondLargest = Number.MIN_SAFE_INTEGER;

  let n = arr.length;
  for (let i = 0; i < n; i++) {
    if (arr[i] > largest) {
      secondLargest = largest; // shift the value of largest into second largest
      largest = arr[i]; // update largest
    } else if (arr[i] > secondLargest && arr[i] != largest) {
      secondLargest = arr[i]; // update second largest
    }
  }
  if (secondLargest == Number.MIN_SAFE_INTEGER) {
    return -1;
  }

  return secondLargest;
}
let arr = [2, 2, 1, 3, 10, 12, 15, 100];
console.log(arr);
console.log("second largest value is : ", secondLargest(arr));
